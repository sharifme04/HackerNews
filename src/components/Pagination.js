import React from 'react';
import PropTypes from 'prop-types';

const Pagination= ({
  storiesPerPage,
  totalStories,
  paginate,
  currentPage,
  previousPage,
  nextPage
}) => {
  let pageNumbers = [];
  for (let i = 1; i <Math.ceil(totalStories/storiesPerPage); i++) {
    pageNumbers = [...pageNumbers, i];
  }
  const intialRange = currentPage-2>= 0 ? currentPage-2: currentPage-1;
  const visiblePagination = pageNumbers.slice(intialRange, currentPage+2);

  return (
  <div>
  <ul className="pagination">
    {currentPage >1 &&
      <li className="page-item">
       <button data-testid="previous" className="page-link" onClick={()=>previousPage(currentPage)}>Previous</button>
      </li>
    }
   {visiblePagination.map(number =>(
     <li key={number} className={`page-item ${ (number === currentPage) && "active"}` }>
      <button data-testid="current" className="page-link" onClick={()=>paginate(number)}>{number}</button>
    </li>
   ))}
   {currentPage < pageNumbers.length  &&
     <li className="page-item">
       <button data-testid="next" className="page-link" onClick={()=>nextPage(currentPage)}>Next</button>
     </li>
   }
  </ul>
 </div>
 );
};

Pagination.propTypes = {
  storiesPerPage: PropTypes.number,
  totalStories: PropTypes.number,
  paginate: PropTypes.func,
  currentPage: PropTypes.number,
  previousPage:PropTypes.func,
  nextPage:PropTypes.func
}

export default Pagination;
