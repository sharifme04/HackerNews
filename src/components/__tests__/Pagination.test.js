import React from "react";
import { shallow } from "enzyme";
import Pagination from "../Pagination";

describe("<Pagination />", () => {
  test("render", () => {
    const component = shallow(
      <Pagination
        storiesPerPage={28}
        totalStories={30}
        paginate={() => {}}
        nextPage={() => {}}
        previousPage={() => {}}
        currentPage={1}
      />
    );
    expect(component).toMatchSnapshot();
  });

  test("return the default when there is no data to map through", () => {
    const component = shallow(<Pagination />);
    expect(component).toMatchSnapshot();
  });

  test("does not break without value", () => {
    const component = shallow(<Pagination />);
    expect(component.find("button")).toHaveLength(0);
  });

  test("does not break with an empty array", () => {
    const component = shallow(<Pagination stories={[]} />);
    expect(component.find("button")).toHaveLength(0);
  });
});
