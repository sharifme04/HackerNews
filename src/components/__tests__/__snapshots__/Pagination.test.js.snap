// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> render 1`] = `
<div>
  <ul
    className="pagination"
  >
    <li
      className="page-item active"
      key="1"
    >
      <button
        className="page-link"
        onClick={[Function]}
      >
        1
      </button>
    </li>
  </ul>
</div>
`;

exports[`<Pagination /> return the default when there is no data to map through 1`] = `
<div>
  <ul
    className="pagination"
  />
</div>
`;
